proposals/0002.md

```diff
@@ -0,0 +1,272 @@
+
+This is a guide to structuring errors well.
+
+# Explanation
+
+## The great `Error` unification
+
+The ultimate goal is to make your errors as easy to integrate and handle as possible.
+The canonical way of achieving this is by implementing the `std::error::Error` trait.
+It's universally used in the Rust ecosystem and provides a common interface for working with
+different errors coming from numerous crates.
+It's not possible to enforce the exact behavior of an implementation of the `Error` trait, but
+there's a widely used convention making it reasonably uniform and predictable.
+
+## The correct behavior of an `Error` implementation
+
+`Error` provides 2 core functionalities:
+
+- it provides the `source` method, which may or may not return a reference to another
+`std::error::Error`
+- it derives from `Display` making all its implementors printable
+
+### The bevavior of `source`
+
+`source` should return a value when the error has another error nested inside.
+The wrapped error should be the cause of the outer one, usually a bit lower level and more detailed.
+The errors can be wrapped multiple times to express the whole chain of errors down to the root
+cause. It can be traversed by calling `source` recursively.
+
+For example the custom `FileLoadError` can wrap the `std::io::Error`.
+This way the file loading function doesn't have to know about all the possible file system failures
+described by the `std::io::Error`.
+When it gets one, it simply wraps it in the `FileLoadError` and returns it.
+If a higher level function is interested, what caused the `FileLoadError`, it can call its `source`
+method and get its cause.
+
+`source` should return `None` when the error is the root cause and doesn't wrap another error.
+It's also impossible to return a wrapped error when it doesn't implement `Error`.
+
+### The bevavior of `Display`
+
+The `Display` implementation should briefly describe the error.
+It should never include information provided by the error returned from the `source`.
+Its output should be formatted as a single line, preferably a single sentence.
+If it's a single sentence, it should start with a lowercase letter and end with no interpunction.
+
+Examples:
+
+`failed to load the file '/tmp/file'` **good**
+
+`failed to load the file '/tmp/file': file not found`
+**bad**: includes information provided by its `source`
+
+`Failed to load the file '/tmp/file'!`
+**bad**: a single sentence starting with a capital letter and ending with interpunction
+
+`Failed to load the file. Path: '/tmp/file'.`
+**ugly**: it has two sentences
+
+## Printing a chain of `Error`s
+
+When all the errors in a chain behave correctly, insightful and well formatted messages can be
+easily provided to the user.
+The error can be iterated by recursively calling `source` to print each layer and provide the full
+chain of errors that caused the failure.
+Due to strict `Display` implementation rules there's a lot of creative freedom when it comes to
```

Comment:
What’s the rules? Seems pretty lawless to me. Or do you mean just the fact that an impl must exist.

Reply:
I've specified that it's about `rules outlined in the previous section`. Is it clear now?